pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/typeinterference/InferenceRuleType.java

```diff
@@ -0,0 +1,263 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.typeresolution.typeinterference;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import net.sourceforge.pmd.lang.java.typeresolution.MethodTypeResolution;
+import net.sourceforge.pmd.lang.java.typeresolution.typedefinition.JavaTypeDefinition;
+
+
+public enum InferenceRuleType {
+
+    /**
+     * https://docs.oracle.com/javase/specs/jls/se8/html/jls-18.html#jls-18.2.4
+     */
+    EQUALITY {
+        @Override
+        public List<BoundOrConstraint> reduce(BoundOrConstraint val) {
+            // A constraint formula of the form ‹S = T›, where S and T are types, is reduced as follows:
+
+            if (val.isLeftType() && val.isRightType()) {
+
+                List<BoundOrConstraint> newConstraints = new ArrayList<>();
+
+                // If S and T are proper types, the constraint reduces to true if S is the same as T (§4.3.4), and false
+                // otherwise.
+                if (val.isLeftProper() && val.isRightProper()) {
+                    if (val.leftProper().equals(val.rightProper())) {
+                        return newConstraints;
+                    } else {
+                        return null;
```

Comment:
That's a tough one... using exceptions for normal / expected behavior is a code smell (exceptions should be exceptional!)... null checking on the other hand is not intuitive and may produce an NPE. `Optional` is not an option since it's Java 8+ only....

We may leave it this way for now, but I'd like us to think on better alternatives